### ISSUE TITLE
Preserving gradients with copy()?

### ISSUE THREAD

**Issue by Gadersd** [LaurentMazare/tch-rs#865]
Does the `copy()` tensor method maintain a gradient back to the source tensor?

I have a tensor that I need to send to multiple `Module`s and used `copy()` to achieve the duplication. Might this cause issues with the gradient or is there a better way to do this?

Also, I intend to use `trainable_variables()` on my `VarStore` to loop through the parameters and compute a regularization loss. Might this be problematic and is there a better way to do this?

**Comment by LaurentMazare** [comment]
`copy` uses PyTorch `copy_` operator behind the hood and from a quick experiment on the python side this seems to properly propagate the gradient through. Did you try it out in rust and that wasn't the case?

```
>>> t = torch.tensor([0.1], requires_grad=True)
>>> u = torch.tensor([0.])
>>> u.copy_(t + 1)
tensor([1.1000], grad_fn=<CopyBackwards>)
>>> u.backward()
>>> t.grad
tensor([1.])
```

Re iterating on the trainable variables, this sounds perfectly fine, weight decay is already handled by most optimizers if that's what you're after but for another regularization loss that would probably be the way to go.
